### PR TITLE
ISPN-8621 Share connection for events

### PR DESCRIPTION
**Pull request by rvansa**
https://issues.jboss.org/browse/ISPN-8621

Set to preview as this depends on #5742

### CODE REVIEW

**Comment by rvansa** [comment]
Removing Preview tag as #5742 was merged.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/impl/ExpiredEventImpl.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+   }
+
+   @Override
+   public Type getType() {
+      return Type.CLIENT_CACHE_ENTRY_EXPIRED;
+   }
+
+   @Override
+   public String toString() {
+      return "ClientCacheEntryExpiredEvent(" + "key=" + key + ")";
```
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/impl/ExpiredEventImpl.java]: Wrong class name now.

Comment (wburns) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/event/impl/ExpiredEventImpl.java]: This seems to be true across all the new events.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/HeaderDecoder.java [hunk 2]
```diff
@@ -62,15 +83,47 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
...
                long messageId = codec.readMessageId(in);
+               receivedOpCode = codec.readOpCode(in);
+               switch (receivedOpCode) {
+                  case CACHE_ENTRY_CREATED_EVENT_RESPONSE:
+                  case CACHE_ENTRY_MODIFIED_EVENT_RESPONSE:
+                  case CACHE_ENTRY_REMOVED_EVENT_RESPONSE:
+                  case CACHE_ENTRY_EXPIRED_EVENT_RESPONSE:
+                     if (codec.allowOperationsAndEvents()) {
+                        operation = messageId == 0 ? null : incomplete.get(messageId);
+                     } else if (incomplete.size() == 1){
```
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/HeaderDecoder.java]: Nitpick: spacing

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/HeaderDecoder.java [hunk 3]
```diff
@@ -91,6 +144,42 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
...
+               break;
+            case READ_COUNTER_EVENT:
+               if (trace) {
+                  log.tracef("Decoding counter event %s", HotRodConstants.Names.of(receivedOpCode));
+               }
+               HotRodCounterEvent counterEvent;
+               try {
+                  counterEvent = codec.readCounterEvent(in);
+               } catch (Signal signal) {
+                  throw signal;
```
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/HeaderDecoder.java]: Seems redundant as the outer catch does this already.

Comment (rvansa) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/HeaderDecoder.java]: This is not redundant, we don't want to log anything for `Signal` in the `catch (Throwable)` below

**Comment by rvansa** [comment]
@wburns Hi, this still requires your attention.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/NotificationManager.java [hunk 4]
```diff
@@ -178,12 +177,10 @@ public void stop() {
       log.debugf("Stopping %s (%s)", this, lock);
       lock.lock();
       try {
-         clientListeners.clear();
-         // Rather than de-registering all the listeners we'll terminate the channel and let server deal with it
-         if (dispatcher != null) {
-            dispatcher.close();
-            dispatcher = null;
-         }
+         CompletableFuture[] futures = clientListeners.keySet().stream().map(counterName ->
```
Comment (wburns) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/NotificationManager.java]: Is it an issue that `clientListeners` isn't cleared any more?

Comment (rvansa) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/impl/NotificationManager.java]: IDK, I have added the clear and the testsuite didn't break...

**Comment by pruivo** [comment]
@rvansa does this PR solves the issues with the channel and events? (see https://issues.jboss.org/browse/ISPN-9053)

**Comment by rvansa** [comment]
@pruivo I recall that I've fixed some bugs introduced in ISPN-8619/ISPN-8620 but I don't recall which ones. Can't say without investigation. Merge it and you'll see.

**Comment by rvansa** [comment]
@wburns @galderz Two approvals is good, merge would be better.

**Comment by wburns** [comment]
Integrated into master, thanks @rvansa 
Sorry was on PTO for a bit there :)

**Pull request merged**